libsolidity/experimental/analysis/FunctionDependencyGraph.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/experimental/analysis/Analysis.h>
+#include <libsolidity/experimental/analysis/FunctionDependencyGraph.h>
+
+using namespace solidity::frontend::experimental;
+using namespace solidity::util;
+
+FunctionDependencyGraph::FunctionDependencyGraph(Analysis& _analysis):
+	m_analysis(_analysis),
+	m_errorReporter(_analysis.errorReporter())
+{
+}
+
+bool FunctionDependencyGraph::analyze(SourceUnit const& _sourceUnit)
+{
+	_sourceUnit.accept(*this);
+	return !m_errorReporter.hasErrors();
+}
+
+bool FunctionDependencyGraph::visit(FunctionDefinition const& _functionDefinition)
+{
+	solAssert(!m_currentFunction);
+	m_currentFunction = &_functionDefinition;
+	return true;
+}
+
+void FunctionDependencyGraph::endVisit(FunctionDefinition const&)
+{
+	// If we're done visiting a function declaration without said function referencing/calling
+	// another function in its body - insert it into the graph without child nodes.
+	annotation().functionCallGraph.edges.try_emplace(m_currentFunction, std::set<FunctionDefinition const*, CallGraph::CompareByID>{});
```

Comment:
A ``solAssert(annotation().functionCallGraph.edges.emplace(m_currentFunction, std::set<FunctionDefinition const*, CallGraph::CompareByID>{}).second);`` in ``visit(FunctionDefinition const& _functionDefinition)`` makes more sense, doesn't it? I.e. checking that it's not already visited and adding the function to the graph when you see it - that's easier to understand than adding it under some special conditions under which it wouldn't be added implicitly elsewhere, isn't it?